### ISSUE TITLE
During parallel execution, memory cost increases on MultiRoom

### ISSUE THREAD

**Issue by Origamizyq** [yfletberliac/adversarially-guided-actor-critic#3]
Hi @yfletberliac
I executed your code  for the environment of "MultiRoom-N10-S10"  with three processes and the default config file (minigrid.yaml).,but the memory usage keeps going up,Could you check it？my plots are as follows.
![Memory](https://user-images.githubusercontent.com/41505001/160743991-90b708b5-8aa1-4027-bbb7-d6968a0ad77b.png)

Thanks,
zhixihuanyu.

**Comment by yfletberliac** [comment]
Hi @yanwangdedongxi,

Thank you for reaching out!

Can I ask what OS Platform and Distribution you are using?
Also, what version do you use for `mpi4py` and `torch`?
And have you used the `requirements.txt` file to install dependencies?

Yannis

**Comment by Origamizyq** [comment]
Hi @yfletberliac

Thank you for your reply.

My OS Platform is 20.04.1-Ubuntu. I used Mpi4py (your code and the version is 3.0.3) for parallel execution and torch framework (your code and  it's version is 1.8.1). 

I have used the file to install dependencies. And I had used your code without parallelism, the memory is stable, the plot is as follow.

Thanks,
zhixihuanyu.
![Memory](https://user-images.githubusercontent.com/41505001/160760871-32195301-faf3-4e55-8b95-2a808506cb63.png)

**Comment by Origamizyq** [comment]
Hi @yfletberliac
 
I'm sorry to ask you some questions again. 

1.Could you tell me the value of "num_episodes_eval" in minigrid.yaml of agac_torch folder, when you made the experiment result of the paper, Is the value 1? 
2.were this plots made from the function named evaluate( ) ?  plots are as follows.
I wish I haven't bothered you.
[
![1648640030(1)](https://user-images.githubusercontent.com/41505001/160825761-4f232f5d-a69a-4ec9-925b-74c4b2d25827.jpg)
]

Thanks,
zhixihuanyu

**Comment by yfletberliac** [comment]
Hi @yanwangdedongxi,

Please find below my answers:

> My OS Platform is 20.04.1-Ubuntu. I used Mpi4py (your code and the version is 3.0.3) for parallel execution and torch framework (your code and it's version is 1.8.1).
> 
> I have used the file to install dependencies. And I had used your code without parallelism, the memory is stable, the plot is as follow.

Sadly it seems that I can't manage to reproduce your issue... You may have also found this but might be interesting: this can also be due to the `numpy` version according to this [issue](https://bitbucket.org/mpi4py/mpi4py/issues/119/memory-leak-allreduce). Although mine is the same as in 'requirements.txt'. Are you also using a GPU? Also, if you let the training continues, does it occur to break at some point because of an out-of-memory error?

> 1. Could you tell me the value of "num_episodes_eval" in minigrid.yaml of agac_torch folder, when you made the experiment result of the paper, Is the value 1?
> 2. Where this plots made from the function named evaluate( ) ?

The average return is computed as the rolling mean over the past 100 episodes. So the plots have simply been made using the data in the training logs.

Yannis

**Comment by Origamizyq** [comment]
Hi @yfletberliac

Thank you for your patience.

I have solved the first problem about memory by using only one process to evaluate the model during parallel execution, thanks for your help.

please allow me to ask this question more accurately, and I hope I didn't bother you. 
- -
"The average return is computed as the rolling mean over the past 100 episodes. So the plots have simply been made using the data in the training logs." 
- -
Did you mean that you used  data of this plot(the plot is as follow)?  
Is the value of "num_episodes_eval"  the default value(the number is 1 in torch framework)?
![1648697212(1)](https://user-images.githubusercontent.com/41505001/160970548-93cf26c2-81ed-402c-8882-47825c7eb173.jpg)

Thanks,
zhixihuanyu

**Comment by yfletberliac** [comment]
Hi @yanwangdedongxi,

No worries I'm happy to answer your questions.

> I have solved the first problem about memory by using only one process to evaluate the model during parallel execution, thanks for your help.

Great!

> Did you mean that you used data of this plot(the plot is as follow)?

I meant that we used the mean returns achieved *during* training and computed an average over the last 100 episodes at each data point in the graph's x-axis. So this would be training_mean_return. But in the case of MiniGrid this is equivalent to using evaluation_mean_return since the episodes are procedurally-generated. So you can use evaluation_mean_return but I would recommend evaluating with at least 10-20 episodes. Or you can use training_mean_return and average over the last 100 episodes each time you generate a point in your graph.
(The reason why the default value is 1 is that the PyTorch version came after the paper was published - we originally used the TF version to produce the paper's graphs, not the PT one.)

Yannis

**Comment by Origamizyq** [comment]
Hi @yfletberliac

Thank you for your answers.

Wish you good health and every success.

Thanks,
zhixihuanyu
